src/shared/math.ts

```diff
@@ -0,0 +1,110 @@
+/** md
+## Math utilities for numberscope
+
+The primary resource for doing math inside the frontscope code is the
+[mathjs](http://mathjs.org) package, which this repository depends on.
+The `shared/math` module of Numberscope provides some additional utilities
+aimed at making common mathematical operations more convenient, especially
+when working with bigints.
+
+Note that currently every one of these functions accepts either
+`number` or `bigint` inputs for all arguments and simply converts them
+to bigints.
+
+### Example usage
+```ts
+import { safeNumber, floorSqrt, modulo, powmod } from '../shared/math'
+try {
+  const myNumber = safeNumber(9007199254740992n)
+} catch (err: unknown) {
+  console.log('This will always be logged')
+}
+try {
+  const myNumber = safeNumber(9007n)
+} catch (err: unknown) {
+  console.log('This never will be and myNumber will be 9007')
+}
+
+const five: bigint = floorSqrt(30n)
+const negativeTwo: bigint = floorSqrt(-2n)
+
+const three: bigint = modulo(-7n, 5)
+const two: bigint = modulo(7, 5n)
+
+const six: bigint = powmod(6, 2401n, 7n)
+// n to the p is congruent to n mod a prime p,
+// so a to any power of p is as well.
+```
+
+### Detailed function reference
+**/
+
+import isqrt from 'bigint-isqrt'
+import {modPow} from 'bigint-mod-arith'
+
+const maxSafeNumber = BigInt(Number.MAX_SAFE_INTEGER)
+const minSafeNumber = BigInt(Number.MIN_SAFE_INTEGER)
+
+/** md
+#### safeNumber(n: number | bigint): number
+
+Returns the `number` mathematically equal to _n_ if there is one, or
+throws an error otherwise.
+**/
+export function safeNumber(n: number | bigint): number {
```

Comment:
Soft suggestion: I like to list exports at the bottom of the module so that you can see all the exports in one place.

Reply:
Hmm, and I like the self-documenting aspect of in-line exports that tells you as you are looking at the function or class that this is one of the deliverables of the module (as opposed to an internal function). I see the benefits of each approach, and unfortunately you can't do both (I tried, and redundant exports are not allowed). How to resolve the disparity in our respective judgment of the pros and cons of these? Most of the code base uses in-line exports, so we could just keep that for consistency, or on Monday we could have a three-way discussion/decision which way to go and if it ends up with bottom exports put an issue in to convert the rest of the codebase...

I will implement the rest of the needed changes a bit later today, but I wanted to give you some time to contemplate this.

Reply:
I guess I can crystallize my discomfort with bottom exports by saying that in my eyes, it increases non-locality of information.

Reply:
Oh I and I think if we decide to prefer in-line exports it should go in the style guide, and if we decide to do bottom-exports we should install and turn on: https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/exports-last.md

Reply:
We're going with in-line exports.